CylindricalEndcap Map (general version)

markscheel opened the pull request:
## Proposed changes

Adds a map from a 3D unit right cylinder to a volume that connects portions of two spherical surfaces.

CylindricalEndcap is intended to be composed with Wedge2D maps to
construct the "endcap" portion of a cylindrical domain for a binary system.

This is a version of #2329 (now closed) that is more general and written in a way that is easier to add similar maps.

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

markscheel committed:
TestMapHelpers: Add test_suite_for_map_on_cylinder.

markscheel committed:
Add Test_CylindricalEndcap.

markscheel committed:
Add new functions to QuadraticEquation.

markscheel commented:
@gsb76 I assigned you to this since you were looking at #2329.  

This PR factors CylindricalEndcap into two classes, one FocallyLiftedMap that handles mapping from an inner surface and a projection point, and another FocallyLiftedEndcap that specializes to a portion of a sphere.

This is currently extremely incomplete and has only documentation (and skeleton hpp files).  I was hoping you could please look at the documentation before I actually wrote the functions.

EDIT: This PR is now complete.

markscheel committed:
Add FocallyLiftedMapHelpers.

markscheel commented:
The only failing test is a macOS test, and the failure occurs in a SWSH subroutine that I did not touch.

markscheel committed:
Add FocallyLiftedMap.

This is a general map that takes a FocallyLiftedInnerMap as
a template argument.

markscheel committed:
Add FocallyLiftedEndcap.

nilsdeppe commented:
@geoffrey4444 can you review this?

nilsdeppe commented:
@geoffrey4444 can you review this?

nilsdeppe commented:
@geoffrey4444 can you review this?

geoffrey4444 commented on src/NumericalAlgorithms/RootFinding/QuadraticEquation.cpp:
```diff
@@ -23,8 +28,139 @@ std::array<double, 2> real_roots(const double a, const double b,
+namespace detail {
+template <>
+struct root_between_values_impl<double> {
+  static double f(const double a, const double b, const double c,
```
name this "function" instead of "f"?

geoffrey4444 commented on src/NumericalAlgorithms/RootFinding/QuadraticEquation.cpp:
```diff
@@ -23,8 +28,139 @@ std::array<double, 2> real_roots(const double a, const double b,
+
+template <>
+struct root_between_values_impl<DataVector> {
+  static DataVector f(const DataVector& a, const DataVector& b,
```
Name function instead of f?

geoffrey4444 commented on src/NumericalAlgorithms/RootFinding/QuadraticEquation.cpp:
```diff
@@ -23,8 +28,139 @@ std::array<double, 2> real_roots(const double a, const double b,
+    // Roots are returned in increasing order.
+    const auto roots = real_roots(a, b, c);
+
+    const std::array<bool, 2> min_bad{
```
Maybe more descriptive names here, like `roots_out_of_bounds_low` and `root_out_of_bounds_high` instead of `min_bad` and `max_bad`?

geoffrey4444 commented on src/NumericalAlgorithms/RootFinding/QuadraticEquation.cpp:
```diff
@@ -23,8 +28,139 @@ std::array<double, 2> real_roots(const double a, const double b,
+      // Check roots[0] first because it is the smallest
+      if (min_bad[0]) {
+        if (min_bad[1]) {
+          ERROR("No root >= min_value.  Roots are "
```
No root >= (within roundoff) min_value

geoffrey4444 commented on src/NumericalAlgorithms/RootFinding/QuadraticEquation.cpp:
```diff
@@ -23,8 +28,139 @@ std::array<double, 2> real_roots(const double a, const double b,
+        }
+        if (max_bad[1]) {
+          ERROR("No root <= max_value.  Roots are "
+                << roots[0] << " and " << roots[1] << ", with a=" << a
```
No root <= (within roundoff) max_value

geoffrey4444 commented on src/NumericalAlgorithms/RootFinding/QuadraticEquation.cpp:
```diff
@@ -23,8 +28,139 @@ std::array<double, 2> real_roots(const double a, const double b,
+      // Check roots[1] first because it is the largest
+      if (max_bad[1]) {
+        if (max_bad[0]) {
+          ERROR("No root <= max_value.  Roots are "
```
<= (within roundoff)

geoffrey4444 commented on src/NumericalAlgorithms/RootFinding/QuadraticEquation.cpp:
```diff
@@ -23,8 +28,139 @@ std::array<double, 2> real_roots(const double a, const double b,
+        return_value = roots[0];
+      } else {
+        if (min_bad[1]) {
+          ERROR("No root >= min_value.  Roots are "
```
<= (within roundoff)

geoffrey4444 commented on src/NumericalAlgorithms/RootFinding/QuadraticEquation.cpp:
```diff
@@ -23,8 +28,139 @@ std::array<double, 2> real_roots(const double a, const double b,
+                << " b=" << b << " c=" << c);
+        }
+        if (max_bad[1]) {
+          ERROR("No root <= max_value.  Roots are "
```
Is this error statement? If you get here, isn't min_bad[0] true, so the 0th root is <= min_value and thus also = max_value?

markscheel replied:
Yes, rewording the error message "no root between min_value and max_value (within roundoff)"

geoffrey4444 commented on src/NumericalAlgorithms/RootFinding/QuadraticEquation.cpp:
```diff
@@ -23,8 +28,139 @@ std::array<double, 2> real_roots(const double a, const double b,
+                << " b=" << b << " c=" << c);
+        }
+        if (min_bad[0]) {
+          ERROR("No root >= min_value.  Roots are "
```
If you get here, must it not be true that root 1 is >= min_value, since it's >= max_value?

markscheel replied:
Yes, rewording the etror message "no root between min_value and max_value (within roundoff)"

geoffrey4444 commented on src/NumericalAlgorithms/RootFinding/QuadraticEquation.hpp:
```diff
@@ -17,6 +17,29 @@
+ * \brief Returns the smallest root of a quadratic equation \f$ax^2 +
+ * bx + c = 0\f$ that is greater than the given value, within roundoff.
+ * \returns A root of a quadratic equation.
+ * \requires That there are two real roots.
```
Doesn't it also require that one of the roots is greater than the given value, since if not, it throws a runtime ERROR?

geoffrey4444 commented on src/NumericalAlgorithms/RootFinding/QuadraticEquation.hpp:
```diff
@@ -17,6 +17,29 @@
+ * \f$ax^2 + bx + c = 0\f$ that is between min_value and max_value,
+ * within roundoff.
+ * \returns A root of a quadratic equation.
+ * \requires That there are two real roots.
```
Doesn't this also require that at least one of the roots is between min_value and max_value, because if not, it gives a runtime error? I wonder if you should warn users that this will give a runtime error on failure

geoffrey4444 commented on tests/Unit/NumericalAlgorithms/RootFinding/Test_QuadraticEquation.cpp:
```diff
@@ -42,6 +44,50 @@
+
+    auto root = smallest_root_greater_than_value_within_roundoff(a, b, c, 0.3);
+    CHECK_ITERABLE_APPROX(expected_root_1, root);
+    root = smallest_root_greater_than_value_within_roundoff(a, b, c, 0.5);
```
What about a check with a value of 0.5 + epsilon, to check the "within roundoff" condition?

geoffrey4444 commented on tests/Unit/NumericalAlgorithms/RootFinding/Test_QuadraticEquation.cpp:
```diff
@@ -42,6 +44,50 @@
+    const auto expected_root_1 = make_with_value<T>(used_for_size, 0.5);
+    const auto expected_root_2 = make_with_value<T>(used_for_size, 5.0);
+
+    auto root = largest_root_between_values_within_roundoff(a, b, c, 0.3, 0.6);
```
Add checks for the cases when one of the roots is epsilon outside the bound?

geoffrey4444 commented on tests/Unit/NumericalAlgorithms/RootFinding/Test_QuadraticEquation.cpp:
```diff
@@ -60,4 +106,9 @@ SPECTRE_TEST_CASE("Unit.Numerical.RootFinding.QuadraticEquation",
+  test_smallest_root_greater_than_value_within_roundoff<double>(1.0);
+  test_smallest_root_greater_than_value_within_roundoff(DataVector(2));
+  test_largest_root_between_values_within_roundoff<double>(1.0);
+  test_largest_root_between_values_within_roundoff(DataVector(2));
```
Add a test that the runtime errors are thrown as expected?

geoffrey4444 reviewed (changes requested):
Reviewed first commit (new functions to QuadraticEquation)

geoffrey4444 commented on src/Domain/CoordinateMaps/FocallyLiftedMap.hpp:
```diff
@@ -0,0 +1,275 @@
+ * \f$\sigma(\bar{x},\bar{y},\bar{z})\f$ defined below; these
+ * functions define the mapping from \f$(\bar{x},\bar{y},\bar{z})\f$
+ * to the 2D surface.
+ *
```
A sketch showing two panels: i) \bar{x}^i, ii) P, x_0^i, x_1^i, and x^i, with an arrow from panel 1 to panel 2, would help clarify what this map is doing. Showing level surfaces \sigma=0 and \sigma=1 would help in the sketch, too

geoffrey4444 commented on src/Domain/CoordinateMaps/FocallyLiftedMap.hpp:
```diff
@@ -0,0 +1,275 @@
+ *
+ * where \f$\lambda\f$ is a scalar factor that depends on \f$x_0^i\f$ and
+ * that can be computed by solving a quadratic equation.  This quadratic
+ * equation is derived by demanding that \f$x_1^i\f$ lies on the sphere:
```
Please explicitly define |x^i|^2 = \delta_{ij} x^i x^j.

geoffrey4444 commented on src/Domain/CoordinateMaps/FocallyLiftedMap.hpp:
```diff
@@ -0,0 +1,275 @@
+ * |P^i + (x_0^i - P^i) \lambda - C^i |^2 - R^2 = 0.
+ * \f}
+ *
+ * Now assume that \f$x_0^i\f$ corresponds to a level surface defined
```
I'm confused by this wording. x_0^i is a single point in 3 space; how can it correspond to a 2D surface made of many points? Do you mean to say it intersects or lies in the level surface?

geoffrey4444 commented on src/Domain/CoordinateMaps/FocallyLiftedMap.hpp:
```diff
@@ -0,0 +1,275 @@
+ * \f}
+ *
+ * Now assume that \f$x_0^i\f$ corresponds to a level surface defined
+ * by some scalar function \f$\sigma(\bar{x},\bar{y},\bar{z})=0\f$,
```
why not \sigma(\bar{x}^i)?

geoffrey4444 commented on src/Domain/CoordinateMaps/FocallyLiftedMap.hpp:
```diff
@@ -0,0 +1,275 @@
+ * Then, once \f$\lambda\f$ has been computed and \f$x_1^i\f$ has
+ * been determined, the map is given by
+ *
+ * \f{align}x^i = x_0^i + (x_1^i - x_0^i) \sigma(\bar{x},\bar{y},\bar{z}).\f}
```
Why not \bar{x}^i instead of \bar{x}, \bar{y}, \bar{z}

geoffrey4444 commented on src/Domain/CoordinateMaps/FocallyLiftedMap.hpp:
```diff
@@ -0,0 +1,275 @@
+ * Differentiating Eq. (1) above yields
+ *
+ * \f[
+ * \frac{\partial x_1^i}{\partial x_0^j} = \lambda +
```
should be `= \lambda \delta^i_j`

geoffrey4444 commented on src/Domain/CoordinateMaps/FocallyLiftedMap.hpp:
```diff
@@ -0,0 +1,275 @@
+ *  (x_0^i - P^i) \frac{\partial \lambda}{\partial x_0^j}.
+ * \f]
+ *
+ * and differentiating Eq. (2) above yields
```
and differentiating Eq. (2) and then inserting Eq. (1) yields

geoffrey4444 commented on src/Domain/CoordinateMaps/FocallyLiftedMap.hpp:
```diff
@@ -0,0 +1,275 @@
+ * + (x_1^i - P^i)(P_i - C_{i})}.
+ * \f}
+ *
+ * The Jacobian can be found by differentiating Eq. (3) above and combining
```
I don't see where Eq. (2) comes into this unnumbered equation. I derived it by differentiating Eq. (3) and using the chain rule, recognizing that x_1^j only depends on \bar{x}^j via its dependence on x_0^j (since \lambda depends on x_0^j). Perhaps the docs here should say that instead?

Also, why is this equation not numbered?

geoffrey4444 commented on src/Domain/CoordinateMaps/FocallyLiftedMap.hpp:
```diff
@@ -0,0 +1,275 @@
+ * + \frac{\partial \sigma}{\partial \bar{x}^j} (x_1^i - x_0^i).
+ * \f]
+ *
+ * The inner map provides the function `deriv_sigma`, which returns
```
What is the "inner map"?

geoffrey4444 commented on src/Domain/CoordinateMaps/FocallyLiftedMap.hpp:
```diff
@@ -0,0 +1,275 @@
+ * ### Inverse map.
+ *
+ * Given \f$x^i\f$, we wish to compute \f$\bar{x}\f$,
+ * \f$\bar{y}\f$, and \f$\bar{z}\f$.
```
Why not \bar{x}^j ?

geoffrey4444 commented on src/Domain/CoordinateMaps/FocallyLiftedMap.hpp:
```diff
@@ -0,0 +1,275 @@
+ * Given \f$x^i\f$, we wish to compute \f$\bar{x}\f$,
+ * \f$\bar{y}\f$, and \f$\bar{z}\f$.
+ *
+ * We first find the coordinates \f$x_0^i\f$ that lie on the 2-surface
```
Again, a sketch here would help this make a lot more sense. Show P, x_0, and x^i joined by a line, and show P, x_1, and x^i joined by a line

geoffrey4444 commented on src/Domain/CoordinateMaps/FocallyLiftedMap.hpp:
```diff
@@ -0,0 +1,275 @@
+ * \f{align} x_0^i = P^i + (x^i - P^i) \tilde{\lambda},\f}
+ *
+ * where \f$\tilde{\lambda}\f$ is a scalar factor that depends on
+ * \f$x^i\f$ and is determined by the inner map.  The inner map
```
Unclear what the "inner map" refers to, but shouldn't this be determined not by the inner map but by its inverse?

geoffrey4444 commented on src/Domain/CoordinateMaps/FocallyLiftedMap.hpp:
```diff
@@ -0,0 +1,275 @@
+ * \f{align}
+ * |P^i + (x_0^i - P^i) \lambda - C^i |^2 - R^2 = 0.
+ * \f}
+ *
```
Please write down here what that quadratic equation actually is (what are a,  b, c)?

geoffrey4444 commented on src/Domain/CoordinateMaps/FocallyLiftedMap.hpp:
```diff
@@ -0,0 +1,275 @@
+ * and are defined such that \f$P\f$, \f$x_1^i\f$, and \f$x^i\f$ are co-linear.
+ * \f$x_1^i\f$ is determined by the equation
+ *
+ * \f{align} x_1^i = P^i + (x^i - P^i) \bar{\lambda},\f}
```
Please provide after this the actual quadratic equation (i.e., what are a, b, c in a x^2 + b x + c = 0)?

geoffrey4444 commented on src/Domain/CoordinateMaps/FocallyLiftedMap.hpp:
```diff
@@ -0,0 +1,275 @@
+ * \f}
+ *
+ * Note that we don't actually need to compute \f$x_1^i\f$. Instead, we
+ * can determine \f$\sigma\f$ by the relation
```
by the relation (obtained by solving Eq. (3) for sigma and then inserting Eqs. (5) and (6) to eliminate x_1^j and x_0^j)

geoffrey4444 commented on src/Domain/CoordinateMaps/FocallyLiftedMap.hpp:
```diff
@@ -0,0 +1,275 @@
+ * can determine \f$\sigma\f$ by the relation
+ *
+ * \f{align}
+ * \sigma = \frac{\tilde{\lambda}-1}{\tilde{\lambda}-\bar{\lambda}}.
```
Are you in trouble if it so happens that \lambda == \bar{\lambda} ?

geoffrey4444 commented on src/Domain/CoordinateMaps/FocallyLiftedMap.hpp:
```diff
@@ -0,0 +1,275 @@
+ * (where \f$\sigma\f$ is held fixed), and a function `dxbar_dsigma`
+ * that returns \f$\partial \bar{x}^i/\partial \sigma\f$ (where
+ * \f$x_0^i\f$ is held fixed).  The factor \f$\partial x_0^j/\partial
+ * x^i\f$ can be computed by differentiating Eq. (5), which yields
```
are you in trouble if \tilde{\lambda} == 0?

geoffrey4444 commented on src/Domain/CoordinateMaps/FocallyLiftedMap.hpp:
```diff
@@ -0,0 +1,275 @@
+ *
+ * To evaluate the remaining unknown factor in Eq. (9),
+ * \f$\partial \sigma/\partial x^j\f$,
+ * note that \f$\bar{\lambda}=\tilde{\lambda}\lambda\f$.
```
note that [combining Eqs. (1), (5), and (6)] \f$\bar{\lambda} = \tilde{\lambda}\lambda\f$

geoffrey4444 commented on src/Domain/CoordinateMaps/FocallyLiftedMap.hpp:
```diff
@@ -0,0 +1,275 @@
+ * \frac{\partial \tilde\lambda}{\partial x^i},
+ * \f}
+ *
+ * where the second factor in the first term can be evaluated using Eq. (4).
```
Eq. (4), the third factor in the first term can be evaluated using Eq. (10), and the second factor in the second term is provided by the inner map via `deriv_lambda_tilde`.

geoffrey4444 commented on src/Domain/CoordinateMaps/FocallyLiftedMapHelpers.hpp:
```diff
@@ -0,0 +1,98 @@
+ * \f]
+ *
+ * This is a quadratic equation for \f$\lambda\f$
+ * and it generally has more than one real root.
```
Please provide the full equation (what are a, b, c in the quadratic you are solving)?

geoffrey4444 commented on src/Domain/CoordinateMaps/FocallyLiftedMapHelpers.hpp:
```diff
@@ -0,0 +1,98 @@
+ * *for all points*, \f$x\f$ is always between src_point
+ * and proj_center. Those maps should set the parameter
+ * `src_is_between_proj_and_target` to false.
+ *
```
Please clarify here that proj_center is P, sphere_center is C, src_point is x_0 (is it?), and intersection_point is x_1 (is it), so you can go between the equations and the inputs to the functions

geoffrey4444 commented on src/Domain/CoordinateMaps/FocallyLiftedMapHelpers.hpp:
```diff
@@ -0,0 +1,98 @@
+ * and proj_center. Those maps should set the parameter
+ * `src_is_between_proj_and_target` to false.
+ *
+ * Note: if we ever add maps where
```
Make this a doxygen warning

geoffrey4444 reviewed (changes requested):
Reviewed documentation for FocallyLiftedMap

geoffrey4444 commented on src/Domain/CoordinateMaps/FocallyLiftedMapHelpers.hpp:
```diff
@@ -0,0 +1,98 @@
+ *  are not as expected (i.e. if the inverse map was called for a
+ *  point not in the range of the map).
+ */
+boost::optional<double> try_scale_factor(
```
Why is this one not templated on type but scale_factor() is?

markscheel replied:
Because this one is needed only for the inverse function, which for all the coordinate maps is only implemented for doubles.

geoffrey4444 commented on src/Domain/CoordinateMaps/FocallyLiftedMapHelpers.hpp:
```diff
@@ -0,0 +1,98 @@
+ * is the quantity returned by `scale_factor`.
+ *
+ * Note that it takes `intersection_point` and not `src_point` as a
+ * parameter.
```
To make the docs here self-contained, please explain or link to docs defining what intersection_point and src_point are?

geoffrey4444 commented on src/Domain/CoordinateMaps/FocallyLiftedMapHelpers.cpp:
```diff
@@ -0,0 +1,215 @@
+    // inside the sphere, 2) src and proj are both outside the sphere,
+    // and 3) proj is outside the sphere and src is inside the sphere.
+    // To cover all 3 cases, we choose the smallest root that is
+    // greater than or equal to unity. This means for case 2) we are
```
Maybe add a clarifying note that a root of 0 corresponds to proj, and a root of 1 corresponds to src, to explain why this choice of selecting the root > unity ensures that src is between proj and the root

geoffrey4444 commented on src/Domain/CoordinateMaps/FocallyLiftedMapHelpers.cpp:
```diff
@@ -0,0 +1,215 @@
+  // and src is outside the sphere 2) src is inside the sphere and proj
+  // is outside the sphere, and 3) the sphere is between src and proj.
+  // To cover all 3 cases, we choose the largest root that is less than
+  // or equal to unity, and we require that this root is positive.
```
A clarifying note that a root of 0 == proj and a root of 1 == src would help clarify why this requirement enforces that the root selected is between proj and src

geoffrey4444 commented on src/Domain/CoordinateMaps/FocallyLiftedMapHelpers.cpp:
```diff
@@ -0,0 +1,215 @@
+    // We solved for scale_factor-1 above, so add 1 to get scale_factor.
+    x0 += 1.0;
+    x1 += 1.0;
+    if (equal_within_roundoff(x0, 1.0)) {
```
UNLIKELY?

geoffrey4444 commented on src/Domain/CoordinateMaps/FocallyLiftedMapHelpers.cpp:
```diff
@@ -0,0 +1,215 @@
+  // happens when inverse-mapping the boundaries.
+
+  // quadratic equation is
+  // a x^2 + b x + c = 0
```
Please add a derivation of these terms to the help text? I checked that they are correct, but it's subtle, because if you forget that you're solving for scale_factor - 1, a is unchanged and b and c are almost unchanged

geoffrey4444 commented on src/Domain/CoordinateMaps/FocallyLiftedMapHelpers.cpp:
```diff
@@ -0,0 +1,215 @@
+    const bool src_is_between_proj_and_target) noexcept {
+  using return_type = tt::remove_cvref_wrap_t<T>;
+  // quadratic equation is
+  // a x^2 + b x + c = 0
```
Here you are not solving for scale_factor - 1 ... why not? Aren't you concerned about the same roundoff issue as in try_scale_factor? 

If you do want to keep both versions, please add a derivation of both to the help text, since they are so similar it's easy for a reader to confuse one for the other

geoffrey4444 commented on src/Domain/CoordinateMaps/FocallyLiftedMapHelpers.cpp:
```diff
@@ -0,0 +1,215 @@
+    const std::array<double, 3>& proj_center,
+    const std::array<double, 3>& sphere_center, const T& lambda) noexcept {
+  using return_type = tt::remove_cvref_wrap_t<T>;
+  const return_type lambda_over_denominator =
```
lambda_squared_over_denominator

geoffrey4444 commented on src/Domain/CoordinateMaps/FocallyLiftedMapHelpers.cpp:
```diff
@@ -0,0 +1,215 @@
+                         (intersection_point[1] - proj_center[1]) *
+                             (proj_center[1] - sphere_center[1]) +
+                         (intersection_point[2] - proj_center[2]) *
+                             (proj_center[2] - sphere_center[2])));
```
Are you in trouble if projection point == sphere center or intersection point == projection point? Could that ever happen?

geoffrey4444 commented on src/Domain/CoordinateMaps/FocallyLiftedMapHelpers.hpp:
```diff
@@ -0,0 +1,98 @@
+    bool pick_larger_root, bool pick_root_greater_than_one) noexcept;
+
+/*!
+ * Computes \f$\partial \lambda/\partial x^i\f$, where \f$\lambda\f$
```
Doesn't it actually compute \partial \lambda / \partial x_0^I?

Please provide the explicit formula here (Eq. (4) in the help text of FocallyLiftedMap.hpp help text)

geoffrey4444 reviewed (changes requested):
Reviewed the FocallyLiftedMap helpers

markscheel commented:
Rebased onto develop, made various fixup commits in response to review by @geoffrey4444

geoffrey4444 commented:
I'm happy with the fixups...the new figure is really helpful, as are the expanded docs! I'll continue reviewing the next commit next

geoffrey4444 commented on src/Domain/CoordinateMaps/FocallyLiftedEndcap.hpp:
```diff
@@ -0,0 +1,342 @@
+ * \details The domain of the map is a 3D unit right cylinder with
+ * coordinates \f$(\bar{x},\bar{y},\bar{z})\f$ such that
+ * \f$-1\leq\bar{z}\leq 1\f$ and \f$\bar{x}^2+\bar{y}^2 \leq
+ * 1\f$.  The range of the map has coordinates \f$(x,y,z)\f$.
```
A figure here would be very helpful for readers understanding what this endcap map is doing

geoffrey4444 commented on src/Domain/CoordinateMaps/FocallyLiftedEndcap.hpp:
```diff
@@ -0,0 +1,342 @@
+ * the 3D coordinates on that sphere, which are given by
+ *
+ * \f{align}
+ * x_0^0 &= R \frac{\sin(\bar{\rho} \theta_\mathrm{max})
```
I'm confused here, because normally I expect the inputs to sin() and cos() to be dimensionless, but here they seem to have dimensions of length. Can you please help me understand what's going on here?

markscheel replied:
rho is dimensionless.  Maybe I could say xbar and ybar lie on a circle of radius Rbar (where Rbar is always 1), and define rho^2 = (xbar^2+ybar^2)/Rbar^2.  This introduces an extra variable Rbar, which is always 1.

geoffrey4444 commented on src/Domain/CoordinateMaps/FocallyLiftedEndcap.hpp:
```diff
@@ -0,0 +1,342 @@
+ * ### sigma
+ *
+ * \f$\sigma\f$ is a function that is zero on the sphere
+ * \f$x^i=x_0^i\f$ and unity at \f$\bar{z}=+1\f$ (corresponding to the
```
Clearer to say sigma is a function that is zero at \bar{z}=-1 (which maps onto the sphere x^I=x_0^i) and unity at \bar{z}=+1 (corresponding to the upper surface of the FocallyLiftedMap)? I suspect the figure I requested above would make this clearer as well

geoffrey4444 commented on src/Domain/CoordinateMaps/FocallyLiftedEndcap.hpp:
```diff
@@ -0,0 +1,342 @@
+ * \f}
+ * where care must be taken to evaluate
+ * \f$\sin(\bar{\rho}\theta_\mathrm{max})/\bar{\rho}\f$
+ * and its derivatives near \f$\bar{\rho}=0\f$.
```
I checked and agree with these equations. Instead of "care must be taken," specify how you do the evaluation? When is close enough that you switch, and what do you switch to?

geoffrey4444 commented on src/Domain/CoordinateMaps/FocallyLiftedEndcap.hpp:
```diff
@@ -0,0 +1,342 @@
+ * \qquad \hbox{(for small $\bar{\rho}$)},
+ * \f}
+ *
+ * where we use the Taylor approximation only for small \f$\bar{\rho}\f$.
```
How small is small enough?

geoffrey4444 commented on src/Domain/CoordinateMaps/FocallyLiftedEndcap.hpp:
```diff
@@ -0,0 +1,342 @@
+ * \f}
+ *
+ * In solving the quadratic, we choose the larger root if
+ * \f$x^2>z_\mathrm{P}\f$ and the smaller root otherwise. We demand
```
Is there a simple explanation you could add here about why this is the way you choose the root?

markscheel commented:
@geoffrey4444 Do you want me to squash my current fixups?

geoffrey4444 commented on src/Domain/CoordinateMaps/FocallyLiftedEndcap.hpp:
```diff
@@ -0,0 +1,342 @@
+ * \left(\bar{\rho} \frac{dq}{d\bar{\rho}} + q\right)^{-1},
+ * \f]
+ *
+ * where \f$\upsilon\f$ is the quantity defined by Eq. (7).  Therefore
```
The LHS has units of 1/length^2, while the RHS has units of 1/length. I think you're missing a 1/R on the RHS of these equations.

geoffrey4444 commented on src/Domain/CoordinateMaps/FocallyLiftedEndcap.hpp:
```diff
@@ -0,0 +1,342 @@
+ * Note that care must be taken to evaluate
+ * \f$q = \sin(\bar{\rho}\theta_\mathrm{max})/\bar{\rho}\f$ and its
+ * derivative near \f$\bar{\rho}=0\f$.
+ *
```
I checked and agree with these equations

geoffrey4444 commented on src/Domain/CoordinateMaps/FocallyLiftedEndcap.hpp:
```diff
@@ -0,0 +1,342 @@
+  Endcap& operator=(Endcap&&) = default;
+
+  template <typename T>
+  std::array<tt::remove_cvref_wrap_t<T>, 3> operator()(
```
Please double-check that these are in the same order as the help text

markscheel replied:
What are 'these'?

nilsdeppe replied:
presumably the functions :shrug:

geoffrey4444 commented on src/Domain/CoordinateMaps/FocallyLiftedEndcap.hpp:
```diff
@@ -0,0 +1,342 @@
+      noexcept;
+
+  template <typename T>
+  tnsr::Ij<tt::remove_cvref_wrap_t<T>, 3, Frame::NoFrame> jacobian(
```
This is probably not something you can change in this PR, but how come the interface wants tensors for Jacobins but arrays for coordinates, instead of tensors for both?

markscheel replied:
Dunno, that's the interface of all the CoordinateMaps.

nilsdeppe replied:
I think partly historical, but we want the coord maps to be frame-agnostic (it is the final CoordinateMap class that assigns the frames) and so we use `std::array`. For Jacobians it seems easier to use actual `Tensor`s and then the `NoFrame` was added. So yes, could be changed. Unclear it's worth the effort. :shrug:

geoffrey4444 commented on src/Domain/CoordinateMaps/FocallyLiftedEndcap.hpp:
```diff
@@ -0,0 +1,342 @@
+  friend bool operator==(const Endcap& lhs, const Endcap& rhs) noexcept;
+  std::array<double, 3> center_{};
+  double radius_{std::numeric_limits<double>::signaling_NaN()};
+  double theta_{std::numeric_limits<double>::signaling_NaN()};
```
Is this thetamax from the help text? If so, why not name it theta_max?

geoffrey4444 commented on src/Domain/CoordinateMaps/FocallyLiftedEndcap.cpp:
```diff
@@ -0,0 +1,297 @@
+
+namespace Endcap_detail {
+double sin_ax_over_x(double x, double ax, double a) noexcept {
+  return square(ax) < 6.0 * std::numeric_limits<double>::epsilon()
```
Where does the 6.0 come from?

wthrowe replied:
`sin(a x)/x`is `a` up to roundoff for roughly `eps < x < sqrt(eps)` either way you evaluate it, so it doesn't really matter.  (The branch is just to prevent division by zero.)  I don't know why `6 eps` was chosen.

geoffrey4444 commented on src/Domain/CoordinateMaps/FocallyLiftedEndcap.cpp:
```diff
@@ -0,0 +1,297 @@
+             ? a
+             : sin(ax) / x;
+}
+double sin_ax_over_x(double x, double a) noexcept {
```
Please comment why you need a 3-argument version and a 2-argument version of this function

geoffrey4444 commented on src/Domain/CoordinateMaps/FocallyLiftedEndcap.cpp:
```diff
@@ -0,0 +1,297 @@
+  return result;
+}
+double dlogx_sin_ax_over_x(double x, double ax, double a) noexcept {
+  return square(ax) < 10.0 * std::numeric_limits<double>::epsilon()
```
How come before you used 6.0 but here you use 10.0?

wthrowe replied:
This one does matter, and should be `pow<4>(a x) < 30 eps`, for maximum error roughly `sqrt(eps/30) a^3 ~ 3e-9 a^3`.  If you keep another term you can use the expansion for `pow<6>(a x) < 840 eps` to get an error of about `cbrt(eps^2/840) a^3 ~ 4e-12 a^3`.

geoffrey4444 commented on src/Domain/CoordinateMaps/FocallyLiftedEndcap.cpp:
```diff
@@ -0,0 +1,297 @@
+      }()),
+      theta_(acos((z_plane - center_[2]) / radius_)) {
+  ASSERT(z_plane != center[2],
+         "Plane must intersect sphere at more than one point");
```
I don't get this error message...maybe just say z_plane must not equal c2? Or otherwise explain why these are equivalent

geoffrey4444 commented on src/Domain/CoordinateMaps/FocallyLiftedEndcap.hpp:
```diff
@@ -0,0 +1,342 @@
+ *
+ * Consider a sphere with center \f$C^i\f$ and radius \f$R\f$ that is
+ * intersected by a plane normal to the \f$z\f$ axis and located at
+ * \f$z = z_\mathrm{P}\f$. `Endcap` provides the following functions:
```
Mention here that you assert that zp != C2

geoffrey4444 commented on src/Domain/CoordinateMaps/FocallyLiftedEndcap.cpp:
```diff
@@ -0,0 +1,297 @@
+  using return_type = tt::remove_cvref_wrap_t<T>;
+  const return_type& xbar = source_coords[0];
+  const return_type& ybar = source_coords[1];
+  const return_type rho = sqrt(square(xbar) + square(ybar));
```
rhobar

geoffrey4444 commented on src/Domain/CoordinateMaps/FocallyLiftedEndcap.cpp:
```diff
@@ -0,0 +1,297 @@
+  using return_type = tt::remove_cvref_wrap_t<T>;
+  const return_type& xbar = source_coords[0];
+  const return_type& ybar = source_coords[1];
+  const return_type rho = sqrt(square(xbar) + square(ybar));
```
rhobar

geoffrey4444 commented on src/Domain/CoordinateMaps/FocallyLiftedEndcap.cpp:
```diff
@@ -0,0 +1,297 @@
+  // dx/dybar
+  get<0, 1>(jacobian_matrix) = d_sin_factor * xbar * ybar;
+  // dy/dxbar
+  get<1, 0>(jacobian_matrix) = d_sin_factor * ybar * xbar;
```
is it faster to avoid the repeat calculation by setting the 1,0 term to the 0,1 term?

geoffrey4444 commented on src/Domain/CoordinateMaps/FocallyLiftedEndcap.cpp:
```diff
@@ -0,0 +1,297 @@
+  const return_type q = Endcap_detail::sin_ax_over_x(rho, theta_);
+  const return_type dlogrho_q = Endcap_detail::dlogx_sin_ax_over_x(rho, theta_);
+  const return_type one_over_r_q = 1.0 / (q * radius_);
+  const return_type tmp =
```
please use a descriptive variable name

geoffrey4444 commented on src/Domain/CoordinateMaps/FocallyLiftedEndcap.cpp:
```diff
@@ -0,0 +1,297 @@
+  using return_type = tt::remove_cvref_wrap_t<T>;
+  const return_type& xbar = source_coords[0];
+  const return_type& ybar = source_coords[1];
+  const return_type rho = sqrt(square(xbar) + square(ybar));
```
rhobar

geoffrey4444 commented on src/Domain/CoordinateMaps/FocallyLiftedEndcap.cpp:
```diff
@@ -0,0 +1,297 @@
+  // for q = sin(rho theta).
+  double sin_rho_theta_over_rho = 0.0;
+  if (square(sin_squared_rho_theta) <
+      (40.0 / 3.0) * std::numeric_limits<double>::epsilon()) {
```
Where does 40.0/3.0 come from?

markscheel replied:
its gone now...

geoffrey4444 commented on src/Domain/CoordinateMaps/FocallyLiftedEndcap.cpp:
```diff
@@ -0,0 +1,297 @@
+
+  // Compute sin^2(rho theta).
+  const double sin_squared_rho_theta = (square(x) + square(y)) / square(r);
+  // Compute sin(rho theta)/rho.
```
I think the sin^2(rho theta) calculation should be a detail function

wthrowe replied:
Don't reinvent the wheel.  Use `boost::math::sinc_pi`.

markscheel replied:
@wthrowe Thanks!  Will use boost::math::sinc_pi.  Do you know if there's a boost function for the derivative of sync_pi (which is also problematic at x=0)?  If not, I need to keep my version.

wthrowe replied:
I do not see one for the derivative, no.

The diagonal components of your Jacobian can be evaluated naively, because the first term approaches 1 and the second term becomes small, including its numerical errors.  The off-diagonal terms I don't see any way to evaluate without a series expansion with better than an absolute error of O(eps), which can be a large relative error (although not large compared with the other components of the matrix).

I see a few other places where the math could be simplified in this class.  I'll go through it all sometime in the next few days.

markscheel replied:
boost::math::sync_pi doesn't work out of the box for DataVectors.  Apparently to get it working one would need to add `std::numeric_limits<DataVector>::bla ` functions.  Or maybe there's a way with boost::math::policies.

markscheel replied:
Also, it seems I can't pass blaze expressions into boost::math::sync_pi.

wthrowe replied:
Your Taylor expansion implementation doesn't work for DataVectors or Blaze expressions either.

But it sounds like you need the expansion of the derivative in any case, so I'm fine with implementing them both the same way.

geoffrey4444 reviewed (changes requested):
I reviewed the CylindricalEndCap map commit

geoffrey4444 commented:
Squashing prior fixups fine with me

wthrowe commented on src/Domain/CoordinateMaps/CylindricalEndcap.cpp:
```diff
@@ -0,0 +1,211 @@
+
+  // Now make sure that alpha < pi/2-theta.
+  // The cone on either side of cone_new_apex is ok, so we use abs below.
+  ASSERT(acos(abs(cos_alpha)) < abs(asin(cos_theta)),
```
This is `square(cos_alpha) + square(cos_theta) > 1`, but you can leave this as-is if you prefer because it is an assertion and those should be clear to humans.

wthrowe commented on src/Domain/CoordinateMaps/FocallyLiftedEndcap.cpp:
```diff
@@ -0,0 +1,297 @@
+  const return_type z = radius_ * cos(rho * theta_) + center_[2];
+  const return_type x = sin_factor * xbar + center_[0];
+  const return_type y = sin_factor * ybar + center_[1];
+  return std::array<return_type, 3>{{std::move(x), std::move(y), std::move(z)}};
```
Not math-related, but x,y,z are const.

wthrowe commented on src/Domain/CoordinateMaps/FocallyLiftedEndcap.cpp:
```diff
@@ -0,0 +1,297 @@
+  // make the map singular.  Also sin_rho_theta_over_rho cannot be
+  // zero unless theta_ (a map parameter) is chosen to be zero, which
+  // also would make the map singular.
+  const double xbar = x / (r * sin_rho_theta_over_rho);
```
```
rho = atan2(sqrt(square(x) + square(y)), z) / theta_
xbar = x * rho / sqrt(square(x) + square(y))
```
These expressions do not have precision problems, but you do have to guard against division by zero.  It doesn't matter much how you do that.  Just using `xbar = ybar = 0` when `x = y = 0` should suffice.

wthrowe commented on src/Domain/CoordinateMaps/FocallyLiftedMap.cpp:
```diff
@@ -0,0 +1,342 @@
+tnsr::Ij<tt::remove_cvref_wrap_t<T>, 3, Frame::NoFrame>
+FocallyLiftedMap<InnerMap>::jacobian(
+    const std::array<T, 3>& source_coords) const noexcept {
+  using return_type = tt::remove_cvref_wrap_t<T>;
```
`ReturnType`, everywhere.  (It's being used as a "real" type, not a metavariable.)

wthrowe commented on src/Domain/CoordinateMaps/FocallyLiftedMap.hpp:
```diff
@@ -0,0 +1,339 @@
+ * the `deriv_lambda_tilde` function of `InnerMap`.  Note that for
+ * nonsingular maps there is no worry that \f$\tilde{\lambda}\f$ is
+ * zero in the denominator of the second term of Eq. (13); if
+ * \f$\tilde{\lambda}=0\f$ then $x_0^i=P^i$ by Eq. (7), and therefore
```
`\f`

wthrowe commented on src/Domain/CoordinateMaps/FocallyLiftedMap.cpp:
```diff
@@ -0,0 +1,342 @@
+    return boost::none;
+  }
+
+  // compute sigma in a roundoff-friendly way.
```
As far as I can tell these expressions have pretty much the same precision properties, but this is fairly hard to analyze and I'm not confident in that result.  (Numerically, it looks like they both have error ~ `eps / max(|lambda_tilde - 1|, |lambda_bar - 1|)`.)  It may depend on exactly how the input variables are calculated.

markscheel replied:
I think it does depend on how input variables are calculated.  In early testing, occasionally the random number tests failed with roundoff before I did this.

wthrowe reviewed (commented):
Review of floating point properties.

markscheel commented:
I squashed the changes that @geoffrey4444 wanted me to squash, and added more fixup commits to address reviews by @geoffrey4444 and @wthrowe

wthrowe commented on src/Domain/CoordinateMaps/FocallyLiftedEndcap.cpp:
```diff
@@ -37,8 +37,49 @@ DataVector sin_ax_over_x(const DataVector& x, double a) noexcept {
 double dlogx_sin_ax_over_x(double x, double ax, double a) noexcept {
-  return square(ax) < 10.0 * std::numeric_limits<double>::epsilon()
-             ? -cube(a) / 3.0
+  // Here we need to worry about roundoff. Note that we can expand
```
Nice discussion!

wthrowe commented on src/Domain/CoordinateMaps/FocallyLiftedEndcap.cpp:
```diff
@@ -0,0 +1,331 @@
+template <typename T>
+std::array<tt::remove_cvref_wrap_t<T>, 3> Endcap::operator()(
+    const std::array<T, 3>& source_coords) const noexcept {
+  using return_type = tt::remove_cvref_wrap_t<T>;
```
You still have a bunch of `return_type`s.  Change them all to `ReturnType`.

wthrowe commented:
Looks good.

geoffrey4444 commented on src/Domain/CoordinateMaps/CylindricalEndcap.hpp:
```diff
@@ -0,0 +1,150 @@
+ * angle \f$2\theta\f$ as shown in the above figure. Construct another
+ * cone, the "invertibility cone", with apex \f$S\f$ chosen such that
+ * the two cones intersect at right angles on the circle; thus the
+ * opening angle of the invertibility cone is \f$2(\pi-\theta)\f$.  A
```
Thanks for making the figures! They are very helpful. However, shouldn't the angle in figure "Allowed region for P" labeled pi - \theta be \pi/2 - theta? Otherwise, the angles in the upper right triangle add up to 3\pi/2.

Similarly, shouldn't the angle of the inevitability cone be \pi - 2\theta?

markscheel replied:
Yes, thanks for catching this!

geoffrey4444 commented on src/Domain/CoordinateMaps/CylindricalEndcap.hpp:
```diff
@@ -0,0 +1,150 @@
+ * \f$P\f$ lies inside the invertibility cone, but not between \f$S\f$
+ * and sphere 1.  Placing the projection point \f$P\f$ to the right of
+ * \f$S\f$ (but inside the invertibility cone) is ok for
+ * invertibility.
```
Do you have a reference you can cite for this, since it's not obvious, and someone might want to look up where this invertibility condition comes from?

markscheel replied:
No I don't know of a reference; I added some extra text explaining why.

geoffrey4444 commented on src/Domain/CoordinateMaps/CylindricalEndcap.hpp:
```diff
@@ -0,0 +1,150 @@
+ *
+ * In addition to invertibility and the two additional restrictions
+ * already mentioned above, we demand a few more restrictions on the
+ * map parameters to simplify the logic for the expected use cases, and
```
Pedantic: no comma before and

geoffrey4444 commented on src/Domain/CoordinateMaps/CylindricalEndcap.hpp:
```diff
@@ -0,0 +1,150 @@
+ * map parameters to simplify the logic for the expected use cases, and
+ * to ensure that jacobians do not get too large. We demand:
+ *
+ * - \f$P\f$ is not too close to the edge of the invertibility cone.
```
What counts as too close?

geoffrey4444 commented on src/Domain/CoordinateMaps/CylindricalEndcap.hpp:
```diff
@@ -0,0 +1,150 @@
+ * - \f$P\f$ is not too close to the edge of the invertibility cone.
+ * - \f$P\f$ is contained in sphere 2.
+ * - \f$P\f$ is to the left of \f$z_\mathrm{P}\f$.
+ * - \f$z_\mathrm{P}\f$ is not too close to the center or the edge of sphere 1.
```
What is too close?

geoffrey4444 commented on src/Domain/CoordinateMaps/CylindricalEndcap.hpp:
```diff
@@ -0,0 +1,150 @@
+ * - \f$P\f$ is to the left of \f$z_\mathrm{P}\f$.
+ * - \f$z_\mathrm{P}\f$ is not too close to the center or the edge of sphere 1.
+ * - If a line segment is drawn between \f$P\f$ and any point on the
+ *   intersection circle, the angle between the line segment and
```
Is the intersection circle the surface where \bar{z}=-1 gets mapped to? If so, probably good to say that. If not, I got confused by the docs...please clarify where the intersection surface is either way?

geoffrey4444 commented on src/Domain/CoordinateMaps/CylindricalEndcap.hpp:
```diff
@@ -0,0 +1,150 @@
+  std::array<tt::remove_cvref_wrap_t<T>, 3> operator()(
+      const std::array<T, 3>& source_coords) const noexcept;
+
+  boost::optional<std::array<double, 3>> inverse(
```
why is this one not templated on type T like operator(), Jacobian, invjacobian?

markscheel replied:
Because the inverse is always done pointwise in all the maps, and in the base class. If it were done with the entire array, we would need a better way than boost::optional to signal whether the inverse succeeds, since it can succeed for some points and not others.

geoffrey4444 reviewed (changes requested):
Reviewed CylindricalEndcap.hpp

geoffrey4444 commented on src/Domain/CoordinateMaps/CylindricalEndcap.cpp:
```diff
@@ -0,0 +1,211 @@
+  // follows:
+  const std::array<double, 3> cone_new_apex = {
+      center_one[0], center_one[1], center_one[2] + radius_one / cos_theta};
+  // Cone_new opens in the -z direction with opening angle 2*(pi/2-theta).
```
Mention that cone_new's apex is "S" on the "Allowed region for P" figure in the hpp help text

geoffrey4444 commented on src/Domain/CoordinateMaps/CylindricalEndcap.cpp:
```diff
@@ -0,0 +1,211 @@
+
+  // Another necessary condition for invertibility is that proj_center
+  // cannot lie between sphere_one and cone_new_apex.
+  const double proj_radius_one = sqrt(square(center_one[0] - proj_center[0]) +
```
I'm being pedantic, but might it be cheaper to compare square_proj_radius_one (so no sort) against square(radius_one)?

geoffrey4444 commented on src/Domain/CoordinateMaps/CylindricalEndcap.cpp:
```diff
@@ -0,0 +1,211 @@
+  // added and more unit tests to test these cases.
+
+  ASSERT(proj_center[2] < z_plane,
+         "CylindricalEndcap: The map hasn't been tested for this "
```
this configuration -> (z component of the projection center) < z_{plane} or something along those lines (spell out the condition)

geoffrey4444 commented on src/Domain/CoordinateMaps/CylindricalEndcap.cpp:
```diff
@@ -0,0 +1,211 @@
+                "jacobians are likely to be large and the map has not been "
+                "tested for these parameters.");
+  ASSERT(
+      abs(cos_theta) >= 0.1,
```
Where does this limit come from? It it the same as spec? Add a comment explaining why this is what you mean by too close here or in the help text or both please

markscheel replied:
Ok, explanation added in help text; basically this value was chosen empirically.  In normal usage, I don't expect that we get anywhere near this limit.  But if we do, I'd rather have an error message than have inaccurate maps that we don't know are inaccurate.

geoffrey4444 commented on src/Domain/CoordinateMaps/CylindricalEndcap.cpp:
```diff
@@ -0,0 +1,211 @@
+         "proj_center is contained inside sphere_two");
+
+  // Check if we are too close to singular.
+  ASSERT(acos(abs(cos_alpha)) < abs(0.95 * asin(cos_theta)),
```
Where does 0.95 come from? Also, this renders the check on line 78 redundant, so why do both checks? If you have a good reason to do both, why not do them one right after the other?

geoffrey4444 reviewed (changes requested):
Partial review of CylindricalEndcap.cpp

markscheel committed:
Add CylindricalEndcap.

markscheel commented:
I squashed all the previous fixups, and added new fixups for yesterday's review by @geoffrey4444

markscheel commented:
Added another trivial fixup commit.

geoffrey4444 commented on tests/Unit/Helpers/Domain/CoordinateMaps/TestMapHelpers.hpp:
```diff
@@ -528,6 +528,106 @@ void test_suite_for_map_on_sphere(
+void test_suite_for_map_on_cylinder(const Map& map,
+                                    const double inner_radius,
+                                    const double outer_radius) noexcept {
+  static_assert(Map::dim == 3, "Works only for a 3d map");
```
Expand error message to say "test_suite_for_map_on_cylinder only works for a 3D map"

geoffrey4444 commented on tests/Unit/Helpers/Domain/CoordinateMaps/TestMapHelpers.hpp:
```diff
@@ -528,6 +528,106 @@ void test_suite_for_map_on_sphere(
+    test_serialization(map_to_test);
+    CHECK_FALSE(map_to_test != map_to_test);
+
+    test_coordinate_map_argument_types(map_to_test,
```
This all looks reasonable, but maybe condense with a loop somehow, since you're just running the same 5 lines over and over again? Maybe store the points in a std::array and then loop over points in the array? Unless there's some performance reason, that seems a bit nicer to me

geoffrey4444 commented on tests/Unit/Domain/CoordinateMaps/Test_CylindricalEndcap.cpp:
```diff
@@ -0,0 +1,218 @@
+           square(center_two[2] - center_one[2]));
+
+  // Pick radius of sphere_one not too small compared to the distance
+  // between the centers.
```
Clarify or refer to docs elsewhere what "too small" means here

geoffrey4444 commented on tests/Unit/Domain/CoordinateMaps/Test_CylindricalEndcap.cpp:
```diff
@@ -0,0 +1,218 @@
+
+    // We will construct two new cones. The first we will call
+    // cone_regular. It is constructed so that cone_regular and
+    // cone_one intersect each other on circle_one at right angles.
```
A figure here would be helpful...maybe refer to a figure in the hpp files somewhere, if an existing figure is suitable?

geoffrey4444 commented on tests/Unit/Domain/CoordinateMaps/Test_CylindricalEndcap.cpp:
```diff
@@ -0,0 +1,218 @@
+      // We may need to try several values of alpha.
+      // When one works, exit.
+      while (true) {
+        // Choose some smaller cone inside of cone_regular.
```
Might you occasionally get time outs if it just picks bad points too often? How likely is that to happen? How many tries on average should this take to succeed?

markscheel replied:
I'm adding a comment saying that failure is rare and on average it should succeed on the first try.  But this particular test will change in #2587 in a way that makes failure even more rare (and easier to compute a probability for).

geoffrey4444 reviewed (changes requested):
Finished looking through the code. I didn't double-check all of the math in how the test chooses proj_center, but I presume that as long as the tests pass, this must be working correctly. Otherwise, I've now taken a pass through every commit. Next I'll look at your fixups

geoffrey4444 reviewed (changes requested):
The fixups look good to me! I'm happy for you to squash

markscheel commented:
Rebased onto current develop. Squashed previous fixups.  Added new fixups.

kidder commented:
@geoffrey4444 @gsb76 please look at

nilsdeppe commented on src/NumericalAlgorithms/RootFinding/QuadraticEquation.cpp:
```diff
@@ -6,13 +6,18 @@
   const auto roots = real_roots(a, b, c);
   ASSERT(roots[0] <= 0.0 and roots[1] >= 0.0,
          "There are two positive roots, " << roots[0] << " and " << roots[1]
-         << ", with a=" << a << " b=" << b << " c=" << c);
```
I think this is a white-space only change. Can you please not format all lines in a file? `git clang-format -f` will format only the lines you touched. You might need to specify a version of clangformat: `git clang-format-10 -f`

nilsdeppe commented on src/NumericalAlgorithms/RootFinding/QuadraticEquation.cpp:
```diff
@@ -23,8 +28,143 @@ std::array<double, 2> real_roots(const double a, const double b,
+namespace detail {
+template <typename T>
+struct root_between_values_impl;
+enum class which_to_choose { min, max };
```
camel case. And maybe elaborate on what `which` is? E.g. `RootToChoose` or `WhichRootToChoose`?

nilsdeppe commented on src/NumericalAlgorithms/RootFinding/QuadraticEquation.cpp:
```diff
@@ -23,8 +28,143 @@ std::array<double, 2> real_roots(const double a, const double b,
+template <typename T>
+T smallest_root_greater_than_value_within_roundoff(const T& a, const T& b,
+                                                   const T& c,
+                                                   double value) noexcept {
```
`const double`

nilsdeppe commented on src/NumericalAlgorithms/RootFinding/QuadraticEquation.cpp:
```diff
@@ -23,8 +28,143 @@ std::array<double, 2> real_roots(const double a, const double b,
+
+template <typename T>
+T largest_root_between_values_within_roundoff(const T& a, const T& b,
+                                              const T& c, double min_value,
```
`const double` 2x

nilsdeppe commented on src/NumericalAlgorithms/RootFinding/QuadraticEquation.cpp:
```diff
@@ -23,8 +28,143 @@ std::array<double, 2> real_roots(const double a, const double b,
+      // Check roots[0] first because it is the smallest
+      if (roots_out_of_bounds_low[0]) {
+        if (roots_out_of_bounds_low[1]) {
+          ERROR("No root >= (within roundoff) min_value.  Roots are "
```
Some of our root finders, like Newton-Raphson from boost throw exceptions to make errors recoverable. Would that be helpful here? (I don't know how it'll be used, so I have no idea)

In either case, the `Roots are...` string comes up a lot so maybe factor that into a lambda?

markscheel replied:
Factored into lambda.   Here the errors are because the user specified certain bounds, so at least for the usage here it doesn't seem helpful to me.  Of course for other usages it may be.

nilsdeppe commented on tests/Unit/NumericalAlgorithms/RootFinding/Test_QuadraticEquation.cpp:
```diff
@@ -60,4 +128,9 @@ SPECTRE_TEST_CASE("Unit.Numerical.RootFinding.QuadraticEquation",
   CHECK(approx(1e8) == small_negative[1]);
+
+  test_smallest_root_greater_than_value_within_roundoff<double>(1.0);
+  test_smallest_root_greater_than_value_within_roundoff(DataVector(2));
```
use size 5 like we normally do

nilsdeppe commented on tests/Unit/NumericalAlgorithms/RootFinding/Test_QuadraticEquation.cpp:
```diff
@@ -60,4 +128,9 @@ SPECTRE_TEST_CASE("Unit.Numerical.RootFinding.QuadraticEquation",
+  test_smallest_root_greater_than_value_within_roundoff<double>(1.0);
+  test_smallest_root_greater_than_value_within_roundoff(DataVector(2));
+  test_largest_root_between_values_within_roundoff<double>(1.0);
+  test_largest_root_between_values_within_roundoff(DataVector(2));
```
size 5

nilsdeppe commented on src/Domain/CoordinateMaps/FocallyLiftedMapHelpers.hpp:
```diff
@@ -0,0 +1,153 @@
+ *  works only on doubles.
+ *
+ */
+boost::optional<double> try_scale_factor(
```
Can this (and the others) use a `std::optional` instead?

markscheel replied:
Yes of course!  When I first submitted this PR we hadn't switched to std::optional yet.

markscheel replied:
...except there is still one boost::optional that will remain (for now), which is in the base class for all the CoordinateMaps.  Issue #2653

nilsdeppe commented on src/Domain/CoordinateMaps/FocallyLiftedMapHelpers.cpp:
```diff
@@ -0,0 +1,257 @@
+      // For the inverse map, we want a scale_factor s such that 0 < s <= 1.
+      // Note that gsl_poly_solve_quadratic returns x0 < x1.
+      // So we have six cases:
+      //  a) x0 < x1 <= 0        ->   error
```
maybe instead of `error` this is `no root`? Unsure

nilsdeppe commented on src/Domain/CoordinateMaps/FocallyLiftedMapHelpers.cpp:
```diff
@@ -0,0 +1,257 @@
+    const std::array<T, 3>& intersection_point,
+    const std::array<double, 3>& proj_center,
+    const std::array<double, 3>& sphere_center, const T& lambda) noexcept {
+  using ReturnType = tt::remove_cvref_wrap_t<T>;
```
Should there be an `ASSERT` requiring `x_a^y != P^i`? In that case the denominator is singular

markscheel replied:
Not really needed because checked elsewhere, but I will add it because you are the second person who has asked about it, and code will be clearer with the assert there.

nilsdeppe commented on src/Domain/CoordinateMaps/FocallyLiftedMapHelpers.hpp:
```diff
@@ -0,0 +1,153 @@
+    const std::array<double, 3>& proj_center,
+    const std::array<double, 3>& sphere_center, const T& lambda) noexcept;
+
+}  // namespace domain::CoordinateMaps::FocallyLiftedMapHelpers
```
Tests?

markscheel replied:
Tests are in Test_CylindricalEndcap, which puts all these classes together.

geoffrey4444 commented on tests/Unit/Domain/CoordinateMaps/Test_CylindricalEndcap.cpp:
```diff
@@ -54,20 +57,24 @@ void test_cylindrical_endcap() {
     // We will construct two new cones. The first we will call
     // cone_regular. It is constructed so that cone_regular and
     // cone_one intersect each other on circle_one at right angles.
+    // cone_regular is
```
typo?

geoffrey4444 reviewed (changes requested):
Just one minor thing then please squash immediately

nilsdeppe commented on src/Domain/CoordinateMaps/FocallyLiftedMap.hpp:
```diff
@@ -0,0 +1,339 @@
+ * \f$(\bar{x},\bar{y},\bar{z})\f$ is uniquely determined by `InnerMap`.
+ * The `inverse` function of `InnerMap` takes \f$x_0^i\f$ and \f$\sigma\f$
+ * as arguments, and returns
+ * \f$(\bar{x},\bar{y},\bar{z})\f$, or boost::none if \f$x_0^i\f$ or
```
ticks around `boost::none` (several times in the dox)

nilsdeppe commented on src/Domain/CoordinateMaps/FocallyLiftedMap.cpp:
```diff
@@ -0,0 +1,342 @@
+  const ReturnType lambda_factor = 1.0 - sigma + lambda * sigma;
+  for (size_t i = 0; i < 3; ++i) {
+    for (size_t j = 0; j < 3; ++j) {
+      jacobian_matrix.get(i, j) += lambda_factor * d_inner.get(i, j);
```
isn't this just `=`?I thought the Jacobian matrix is zero so far?

nilsdeppe commented on src/Domain/CoordinateMaps/FocallyLiftedMap.cpp:
```diff
@@ -0,0 +1,342 @@
+      make_with_value<tnsr::Ij<tt::remove_cvref_wrap_t<T>, 3, Frame::NoFrame>>(
+          dereference_wrapper(source_coords[0]), 0.0);
+
+  // Do the easiest of the terms involving the inner map.
```
I don't understand what equation this maps to in the documentation.

nilsdeppe commented on src/Domain/CoordinateMaps/FocallyLiftedMap.cpp:
```diff
@@ -0,0 +1,342 @@
+  // Do the easiest of the terms involving the inner map.
+  const ReturnType sigma = inner_map_.sigma(source_coords);
+  const auto d_inner = inner_map_.jacobian(source_coords);
+  const ReturnType lambda_factor = 1.0 - sigma + lambda * sigma;
```
I don't understand why this isn't `1 - sigma`. From Eq. (6) I would expect it to be `1-sigma`. I think if you substituted Eq. 4 into Eq. 6 and then factored things appropriately this would be clearer

nilsdeppe commented on src/Domain/CoordinateMaps/FocallyLiftedMap.cpp:
```diff
@@ -0,0 +1,342 @@
+    }
+  }
+
+  // Do lambda term, which is the most complicated one.
```
maybe `Do lambda derivative term`? Because it's not like the other terms don't have a lambda in them :)

nilsdeppe commented on src/Domain/CoordinateMaps/FocallyLiftedMap.cpp:
```diff
@@ -0,0 +1,342 @@
+      FocallyLiftedMapHelpers::d_scale_factor_d_src_point<ReturnType>(
+          upper_coords, proj_center_, center_, lambda);
+  for (size_t j = 0; j < 3; ++j) {
+    auto temp = make_with_value<ReturnType>(sigma, 0.0);
```
make the `temp` outside the loop so you don't re-allocate memory, then split the first iteration of the for loop out. I would advocate for not having the for loop at all so we can get more vectorization, but that's speculation on performance :shrug: Even better would be re-using some of your previous allocations.

nilsdeppe commented on src/Domain/CoordinateMaps/FocallyLiftedMap.cpp:
```diff
@@ -0,0 +1,342 @@
+template <typename InnerMap>
+template <typename T>
+tnsr::Ij<tt::remove_cvref_wrap_t<T>, 3, Frame::NoFrame>
+FocallyLiftedMap<InnerMap>::jacobian(
```
My biggest concern with this function is the sheer number of allocations it does internally. I realize some of them can't be helped, but there are several that could be re-used.

nilsdeppe commented on src/Domain/CoordinateMaps/FocallyLiftedMap.cpp:
```diff
@@ -0,0 +1,342 @@
+      InnerMap::projection_source_is_between_focus_and_target());
+
+  // upper_coords are the mapped coords on the surface of the sphere.
+  std::array<ReturnType, 3> upper_coords = lower_coords;
```
I don't understand this assignment... Why not default construct?

nilsdeppe commented on src/Domain/CoordinateMaps/FocallyLiftedMap.cpp:
```diff
@@ -0,0 +1,342 @@
+      make_with_value<tnsr::i<tt::remove_cvref_wrap_t<T>, 3, Frame::NoFrame>>(
+          sigma, 0.0);
+  for (size_t i = 0; i < 3; ++i) {
+    auto tmp = make_with_value<ReturnType>(lambda, 0.0);
```
move `tmp` out of loop, try to re-use allocation

nilsdeppe commented on src/Domain/CoordinateMaps/FocallyLiftedMap.cpp:
```diff
@@ -0,0 +1,342 @@
+    for (size_t j = 0; j < 3; ++j) {
+      for (size_t k = 0; k < 3; ++k) {
+        inv_jacobian_matrix.get(i, j) +=
+            dxbar_dx_inner.get(i, k) * dx0_dx.get(k, j);
```
why call dx0 dx_inner in one place but dx0 in the other?

nilsdeppe commented on src/Domain/CoordinateMaps/FocallyLiftedMap.hpp:
```diff
@@ -0,0 +1,339 @@
+template <typename InnerMap>
+bool operator!=(const FocallyLiftedMap<InnerMap>& lhs,
+                const FocallyLiftedMap<InnerMap>& rhs) noexcept;
+}  // namespace domain::CoordinateMaps
```
This commit has no tests for `FocallyLiftedMap`

markscheel replied:
It is tested thru Test_CylindricalEndcap, whose member functions are basically forwarded to those of FocallyLiftedMap.

nilsdeppe commented on src/Domain/CoordinateMaps/FocallyLiftedEndcap.cpp:
```diff
@@ -0,0 +1,331 @@
+  // we need to treat only that case as special.
+  return x == 0.0 ? a : sin(a * x) / x;
+}
+DataVector sin_ax_over_x(const DataVector& x, double a) noexcept {
```
Why not combine the two functions for `double` and `DataVector`?

markscheel replied:
I could template on type, but how to check `if (x==0)` for `DataVector`?  Especially since some components of x may be zero and others are not.

nilsdeppe commented on src/Domain/CoordinateMaps/FocallyLiftedEndcap.cpp:
```diff
@@ -0,0 +1,331 @@
+  // is 3*eps/(ax)^2 where eps is machine epsilon.  (That expression comes
+  // from replacing "1 - 1" with eps and noting that the correct answer
+  // is the 2(ax)^2/3! term and eps is an error contribution).
+  // This means the error 100% if (ax)^2 is 3*eps.
```
this sentence is not complete. I don't understand what you mean, sorry

nilsdeppe commented on src/Domain/CoordinateMaps/FocallyLiftedEndcap.cpp:
```diff
@@ -0,0 +1,331 @@
+  // error = 3 * [eps^n*((2n+2)/(2n+3)!)]^{1/(n+1)}.
+  //
+  // For certain values of n:
+  // n=1    error=3*sqrt(eps/30)               ~ 5e-9
```
I really appreciate this table :)

nilsdeppe commented on src/Domain/CoordinateMaps/FocallyLiftedEndcap.cpp:
```diff
@@ -0,0 +1,331 @@
+                   (1.0 + square(ax) * (-0.1 + square(ax) / 280.0))
+             : (a * cos(ax) - sin(ax) / x) / square(x);
+}
+DataVector dlogx_sin_ax_over_x(const DataVector& x, double a) noexcept {
```
Might be able to combine the `double` and `DataVector` functions here too, but less obvious

nilsdeppe commented on src/Domain/CoordinateMaps/FocallyLiftedEndcap.cpp:
```diff
@@ -0,0 +1,331 @@
+}
+}  // namespace Endcap_detail
+
+Endcap::Endcap(const std::array<double, 3>& center, double radius,
```
More const

nilsdeppe commented on src/Domain/CoordinateMaps/FocallyLiftedEndcap.cpp:
```diff
@@ -0,0 +1,331 @@
+Endcap::Endcap(const std::array<double, 3>& center, double radius,
+               double z_plane) noexcept
+    : center_(center),
+      radius_([&]() noexcept {
```
capture what you need

nilsdeppe commented on src/Domain/CoordinateMaps/FocallyLiftedEndcap.cpp:
```diff
@@ -0,0 +1,331 @@
+               "Cannot have zero radius");
+        return radius;
+      }()),
+      theta_max_([&]() noexcept {
```
capture what you need

nilsdeppe commented on src/Domain/CoordinateMaps/FocallyLiftedEndcap.hpp:
```diff
@@ -0,0 +1,426 @@
+ * (This expression comes from replacing "1 - 1" with \f$\epsilon\f$
+ * and comparing the lowest-order contribution to the correct answer, i.e.
+ * the \f$2(ax)^2/3!\f$ term, with \f$\epsilon\f$, the error contribution).
+ * This means the error
```
I also don't understand this sentence here :) (same as the cpp file)

nilsdeppe commented on src/Domain/CoordinateMaps/FocallyLiftedEndcap.cpp:
```diff
@@ -0,0 +1,331 @@
+  const ReturnType rhobar = sqrt(square(xbar) + square(ybar));
+  const ReturnType sin_factor =
+      radius_ * Endcap_detail::sin_ax_over_x(rhobar, theta_max_);
+  ReturnType z = radius_ * cos(rhobar * theta_max_) + center_[2];
```
Why not `x, y,` and then `z`? This cause be to start searching for some sort of optimization. Funny enough, there could be two (this brings you down to zero extra allocations):
```
ReturnType z = sqrt(square(xbar) + square(ybar));
ReturnType y = Endcap_detail::sin_ax_over_x(z, theta_max_); 
y *= radius_;
ReturnType x = y * xbar + center_[0];
ReturnType y = y * ybar + center_[1];
z = radius_ * cos(z * theta_max_) + center_[2];
```

nilsdeppe commented on src/Domain/CoordinateMaps/FocallyLiftedEndcap.cpp:
```diff
@@ -0,0 +1,331 @@
+  using ReturnType = tt::remove_cvref_wrap_t<T>;
+  const ReturnType& xbar = source_coords[0];
+  const ReturnType& ybar = source_coords[1];
+  const ReturnType rhobar = sqrt(square(xbar) + square(ybar));
```
I'll let you figure out how to play the allocation-reducing game here :)
You might also be able to reduce the number of FLOPs at the same time, I think

nilsdeppe commented on src/Domain/CoordinateMaps/FocallyLiftedEndcap.cpp:
```diff
@@ -0,0 +1,331 @@
+  using ReturnType = tt::remove_cvref_wrap_t<T>;
+  const ReturnType& xbar = source_coords[0];
+  const ReturnType& ybar = source_coords[1];
+  const ReturnType rhobar = sqrt(square(xbar) + square(ybar));
```
I'm not sure if you'll be able to get rid of _all_ the extra allocations, but I would guess at least some of them

nilsdeppe commented on src/Domain/CoordinateMaps/FocallyLiftedEndcap.hpp:
```diff
@@ -0,0 +1,426 @@
+ * \f{align*}
+ * \frac{\partial \bar{x}}{\partial x_0^0} &=
+ * \frac{1}{R q}
+ * - \frac{\bar{x}^2}{R q \bar{\rho}} \frac{dq}{d\bar{\rho}}
```
The code use `d/dlog(rho)`, which makes it difficult to compare to the documentation. Given the lack of tests, this doesn't give me much confidence in the code

markscheel replied:
I can put in explicit logs in the docs.  Edit: Actually, I got rid of the logs (in dox and code) since everything was clearer without them and I found a mistake in the docs for them which I fixed.

nilsdeppe commented on src/Domain/CoordinateMaps/FocallyLiftedEndcap.hpp:
```diff
@@ -0,0 +1,426 @@
+      noexcept;
+
+  template <typename T>
+  std::array<tt::remove_cvref_wrap_t<T>, 3> deriv_sigma(
```
Maybe I missed it, but it would be good to document why none of the `sigma`, `deriv_sigma`, etc. functions are used anywhere in this class

nilsdeppe commented on src/Domain/CoordinateMaps/FocallyLiftedEndcap.cpp:
```diff
@@ -0,0 +1,331 @@
+                                    2.0)}};
+}
+
+boost::optional<double> Endcap::lambda_tilde(
```
[picking random spot] I agree with @geoffrey4444 that having the order of the functions in the documentation and in the code be consistent would make the code easier to review and understand

nilsdeppe commented on src/Domain/CoordinateMaps/FocallyLiftedEndcap.cpp:
```diff
@@ -0,0 +1,331 @@
+}
+
+template <typename T>
+tt::remove_cvref_wrap_t<T> Endcap::sigma(
```
I would recommend switch all these internal helpers to returning by `not_null`. I'm worried that all the memory allocations will mean someone else will need to do that anyway in the future, plus you might be able to eliminate some allocations immediately.

nilsdeppe commented on src/Domain/CoordinateMaps/FocallyLiftedEndcap.hpp:
```diff
@@ -0,0 +1,426 @@
+  double theta_max_{std::numeric_limits<double>::signaling_NaN()};
+};
+bool operator!=(const Endcap& lhs, const Endcap& rhs) noexcept;
+}  // namespace domain::CoordinateMaps::FocallyLiftedInnerMaps
```
This also appears to be missing tests completely

nilsdeppe commented on src/Domain/CoordinateMaps/CylindricalEndcap.hpp:
```diff
@@ -0,0 +1,162 @@
+ * construct a portion of a cylindrical domain for a binary system.
+ *
+ * CylindricalEndcap is described briefly in the Appendix of
+ * Phys. Rev. D 86, 084033 (2012), https://arxiv.org/abs/1206.3015.
```
please add the paper to the bibliography and use Doxygen's `\cite` command

nilsdeppe commented on src/Domain/CoordinateMaps/CylindricalEndcap.hpp:
```diff
@@ -0,0 +1,162 @@
+ * Figure 20 of that paper.
+ *
+ * CylindricalEndcap is implemented using `FocallyLiftedMap`
+ * and `FocallyLiftedEndcap`; see those classes for details.
```
`FocallyLiftedEndcap` link not working. Wrong name?

nilsdeppe commented on src/Domain/CoordinateMaps/CylindricalEndcap.cpp:
```diff
@@ -0,0 +1,224 @@
+
+  // Now consider a different cone, cone_new, constructed so that
+  // cone_new and cone_one intersect each other on circle_one at right
+  // angles.  The apex of cone_new is cone_new_apex, defined as
```
maybe map these names to the names in the Dox? I think `cone_new_apex` is `S`?

markscheel replied:
Already done a few lines below...

nilsdeppe commented on src/Domain/CoordinateMaps/CylindricalEndcap.cpp:
```diff
@@ -0,0 +1,224 @@
+  // follows:
+  const std::array<double, 3> cone_new_apex = {
+      center_one[0], center_one[1], center_one[2] + radius_one / cos_theta};
+  // Cone_new opens in the -z direction with opening angle 2*(pi/2-theta).
```
If it's not too difficult to do, could you add a little coordinate axis symbol into the corner of the second diagram, "Allowed region for P" (or at least which way the z-axis is)

nilsdeppe commented on src/Domain/CoordinateMaps/CylindricalEndcap.cpp:
```diff
@@ -0,0 +1,224 @@
+  // has an opening angle 2*theta.  Call the cone 'cone_one'.
+  const double cos_theta = (z_plane - center_one[2]) / radius_one;
+
+  // Now consider a different cone, cone_new, constructed so that
```
would `cone_2` be a reasonable name? Could be used in the documentation too

nilsdeppe commented on src/Domain/CoordinateMaps/CylindricalEndcap.cpp:
```diff
@@ -0,0 +1,224 @@
+      "CylindricalEndcap: z_plane is too close to the center of sphere_one. "
+          << "cos_theta = " << cos_theta
+          << ". The map is not singular, but the jacobians are likely to be "
+             "large "
```
[nitpick] formatting: combine this and next line together

nilsdeppe commented on src/Domain/CoordinateMaps/CylindricalEndcap.cpp:
```diff
@@ -0,0 +1,224 @@
+  // are hard limits that shouldn't be touched.
+
+  ASSERT(proj_center[2] < z_plane,
+         "CylindricalEndcap: The map hasn't been tested for the "
```
The function name, file, and line number gets printed as part of the ASSERT so I don't think the `CylindricalEndcap:` prefix is needed.

nilsdeppe commented on src/Domain/CoordinateMaps/CylindricalEndcap.hpp:
```diff
@@ -0,0 +1,162 @@
+ *
+ * - \f$P\f$ is not too close to the edge of the invertibility cone.
+ *   Here we demand that the angle between \f$P\f$ and \f$S\f$
+ *   is less than \f$0.95*(\pi/2-\theta)\f$ (note that if this
```
remove `*` from equation

nilsdeppe commented on src/Domain/CoordinateMaps/CylindricalEndcap.cpp:
```diff
@@ -0,0 +1,224 @@
+  ASSERT(proj_center[2] < max_proj_center_z,
+         "CylindricalEndcap: proj_center is too close to z_plane. The "
+         "map has not been tested for this case.");
+  const double tan_beta = sqrt(square(center_one[0] - proj_center[0]) +
```
I don't know what `beta` is

nilsdeppe reviewed (changes requested):
I still need to review:
- TestMapHelpers: Add test_suite_for_map_on_cylinder.
- Add Test_CylindricalEndcap.
and their fixups. I'll wait until you squash before reviewing those

markscheel committed:
Add new paper to References.bib

markscheel commented:
Squashed all fixups from review of @geoffrey4444.
Added new fixups to address review of @nilsdeppe ; did everything except haven't reordered functions yet cuz I want to avoid conflicts.
Note that I have some fixups that reduce allocations before my fixups to change functions to return-by-not-null and I reduced allocations even further after my fixups to change functions to return-by-not-null.

Also note that Test_CylindricalEndcap.cpp will be simplified/rewritten in  #2587 to accomodate sphere2>sphere1 in that PR (because in that PR the criteria for valid parameters changes).

kidder commented:
@geoffrey4444 @nilsdeppe please look at fixups

kidder commented:
@geoffrey4444 @nilsdeppe please look at fixups

nilsdeppe commented on src/Domain/CoordinateMaps/FocallyLiftedEndcap.cpp:
```diff
@@ -71,36 +74,36 @@ double dlogx_sin_ax_over_x(double x, double a) noexcept {
              : (a * cos(ax) - sin(ax) / x) / square(x);
 }
-DataVector dlogx_sin_ax_over_x(const DataVector& x, double a) noexcept {
+DataVector one_over_x_d_sin_ax_over_x(const DataVector& x, double a) noexcept {
```
`const double a`?

nilsdeppe commented on tests/Unit/Helpers/Domain/CoordinateMaps/TestMapHelpers.hpp:
```diff
@@ -528,6 +516,105 @@ void test_suite_for_map_on_sphere(
+   * \brief Given a Map `map`, tests the map functions, including map inverse,
+   * jacobian, and inverse jacobian, for a series of points.
+   * These points are chosen in a right cylinder with cylindrical radius 1
+   * and I1-axis extending from -1 to +1. The
```
sorry, `I1-axis` here is the "z-axis"? I'm hoping we can avoid some (much) of the SpEC terminology like `I1` `S2`, `B3` etc. when it's at least reasonably easy to do so :)

nilsdeppe commented on tests/Unit/Domain/CoordinateMaps/Test_CylindricalEndcap.cpp:
```diff
@@ -0,0 +1,224 @@
+  std::uniform_real_distribution<> interval_dis(-1.0, 1.0);
+  std::uniform_real_distribution<> angle_dis(0.0, 2.0 * M_PI);
+
+  // Choose some random centers for sphere_one and sphere_two
```
200 lines to figure out the arguments to the constructor :joy:

markscheel replied:
Yeah, the restrictions on the arguments are complicated. #2587 will simplify this.

nilsdeppe commented on tests/Unit/Domain/CoordinateMaps/Test_CylindricalEndcap.cpp:
```diff
@@ -0,0 +1,224 @@
+        } else if (num_real_roots == 1) {
+          r_max = x0;
+        } else {
+          ASSERT(false, "No roots were found. Something is horribly wrong.");
```
any chance I could convince you to give some hint (in the ASSERT message) as to what is "horribly wrong" and where I should look to fix it? :D

nilsdeppe reviewed (commented):
Okay, here's the rest of the review... Please go ahead and squash away anything and everything that you want to squash away, including any new fixups

markscheel commented:
Made changes, squashed, rebased on develop.

nilsdeppe commented:
@geoffrey4444 @wthrowe could you please take a look at this again when you get a chance? Thanks!
